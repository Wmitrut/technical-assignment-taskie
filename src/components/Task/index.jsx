import { useEffect, useState } from 'react';
import { Outlet, useNavigate } from 'react-router-dom';
import { FiShare, FiX } from 'react-icons/fi';
import useFlashMessage from '../../hooks/useFlashMessage';
import {
  Backdrop,
  Container,
  TaskID,
  TaskDescription,
  TaskHeader,
  TaskTitle,
  CardButton,
  CloseButton,
  Copied,
} from './styles';

const Task = (props) => {
  const [copied, setCopied] = useState(false);

  const [flashCopied, setflashCopied] = useFlashMessage();
  const navigate = useNavigate();

  async function copy() {
    await navigator.clipboard.writeText(location.href);
    setCopied(true);
  }

  useEffect(() => {
    if (copied === true) {
      setflashCopied(true);
    }
  }, [copied]);

  return (
    <Backdrop
      onClick={(event) => {
        props.setModalShown(false);
        event.stopPropagation();
        event.preventDefault();
        navigate(`/`);
      }}>
      <Container onClick={(event) => event.stopPropagation()}>
        <TaskHeader>
          <TaskTitle>{props.task.title}</TaskTitle>
          <div style={{ display: 'flex' }}>
            <CardButton onClick={() => copy()} type="button">
              <FiShare />
            </CardButton>
            <CloseButton
              type="button"
              onClick={() => {
                props.setModalShown(false);
                navigate(`/`);
              }}>
              <FiX />
            </CloseButton>
          </div>
        </TaskHeader>

        <TaskID>#{props.task.id}</TaskID>
        <TaskDescription>{props.task.description}</TaskDescription>
        {flashCopied ? <Copied>Card URL copied to clipboard</Copied> : null}
      </Container>
      <Outlet />
    </Backdrop>
  );
};

export default Task;
