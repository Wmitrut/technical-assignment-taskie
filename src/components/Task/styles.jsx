import styled from 'styled-components';
import { COLORS } from '../../constants/colors';

export const Backdrop = styled.div`
  width: 100%;
  height: 100vh;
  background: rgba(0, 0, 0, 0.5);
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: absolute;
  z-index: 9;
  top: 0;
  left: 0;
`;

export const Container = styled.div`
  width: 500px;
  min-height: 150px;
  background-color: ${COLORS.primary};
  color: ${COLORS.light};
  border-radius: 20px;
  padding: 10px 20px;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
`;

export const TaskHeader = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

export const TaskTitle = styled.h3`
  color: ${COLORS.light};
  margin: 2px;
`;

export const TaskID = styled.p`
  color: ${COLORS.light};
  margin: 2px;
  font-size: 12px;
  opacity: 0.4;
`;

export const TaskDescription = styled.p``;

export const CardButton = styled.button`
  width: 28px;
  height: 28px;
  border-radius: 5px;
  border: 1px solid;
  border-color: ${COLORS.secondary};
  background-color: ${COLORS.primary};
  transition: 0.3s all;
  cursor: pointer;
  color: ${COLORS.light};
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: 2px;
  &:hover {
    background: ${COLORS.successDisabled};
  }
`;

export const CloseButton = styled(CardButton)`
  &:hover {
    background: ${COLORS.error};
  }
`;

export const Copied = styled.span`
  background: #4aa3df;
  padding: 5px;
  box-sizing: border-box;
  border-radius: 5px;
  font-size: 12px;
  text-align: center;
  margin: 10px auto;
`;
