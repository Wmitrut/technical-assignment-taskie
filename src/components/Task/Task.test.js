import renderer from 'react-test-renderer';

import Task from './index';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

const task = {
  id: 'task-51662',
  title: 'Feed the cat',
  description: '',
};

it('Renders open task correctly with data', () => {
  const tree = renderer.create(<Task task={task} />).toJSON();
  expect(tree).toMatchSnapshot();
});
