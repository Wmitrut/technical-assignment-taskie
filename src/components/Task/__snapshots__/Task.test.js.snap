// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders open task correctly with data 1`] = `
<div
  className="sc-bczRLJ kUPrQH"
  onClick={[Function]}
>
  <div
    className="sc-gsnTZi hXTvva"
    onClick={[Function]}
  >
    <div
      className="sc-dkzDqf oOXro"
    >
      <h3
        className="sc-hKMtZM lhzFP"
      >
        Feed the cat
      </h3>
      <div
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <button
          className="sc-gKXOVf fWtExf"
          onClick={[Function]}
          type="button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 12v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2v-8"
            />
            <polyline
              points="16 6 12 2 8 6"
            />
            <line
              x1="12"
              x2="12"
              y1="2"
              y2="15"
            />
          </svg>
        </button>
        <button
          className="sc-gKXOVf sc-iBkjds fWtExf ihjOmI"
          onClick={[Function]}
          type="button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
      </div>
    </div>
    <p
      className="sc-eCYdqJ cCXUtV"
    >
      #
      task-51662
    </p>
    <p
      className="sc-jSMfEi"
    />
  </div>
</div>
`;
